server/src/platform/drbg/helpers.rs

```diff
@@ -0,0 +1,52 @@
+use super::Drbg;
+
+/// Implement `fill_bytes` via `next_u64` and `next_u32`, little-endian order.
+///
+/// The fastest way to fill a slice is usually to work as long as possible with
+/// integers. That is why this method mostly uses `next_u64`, and only when
+/// there are 4 or less bytes remaining at the end of the slice it uses
+/// `next_u32` once.
+pub fn fill_bytes_via_next<R: Drbg>(
+    rng: &mut R,
+    additional_input: &[u8],
+    dest: &mut [u8],
+) -> Result<(), R::Error> {
+    let mut left = dest;
+    while left.len() >= 8 {
+        let (l, r) = { left }.split_at_mut(8);
+        left = r;
+        let chunk: [u8; 8] = rng.next_u64(additional_input)?.to_le_bytes();
+        l.copy_from_slice(&chunk);
+    }
+    let n = left.len();
+    if n > 4 {
+        let chunk: [u8; 8] = rng.next_u64(additional_input)?.to_le_bytes();
+        left.copy_from_slice(&chunk[..n]);
+    } else if n > 0 {
+        let chunk: [u8; 4] = rng.next_u32(additional_input)?.to_le_bytes();
+        left.copy_from_slice(&chunk[..n]);
+    }
+    Ok(())
+}
+
+/// Implement `next_u32` via `fill_bytes`, little-endian order.
+pub fn next_u32_via_fill<R: Drbg>(rng: &mut R, additional_input: &[u8]) -> Result<u32, R::Error> {
```

Comment:
I'm not totally sure why these are needed (all functions below this point), given that it seems the primitives of our `Drbg` trait are `next_u32` and `next_u64` respectively. It seems the only place you're using these are in tests? I would not provide them in a non-test helper module unless we see that we actually need to provide these in a public API sense.

And personally, I'd implement `fill_bytes_via_next` as a default implementation on the trait, instead of providing it as a separate helper. So really, I'd basically get rid of this module.

Reply:
So the idea is that a drbg might have hardware acceleration for any combination of the following:
- `next_u32`
- `next_u64`
- `fill_bytes`

if all three are provided as hardware acceleration, then none of the helpers are needed. If only one is provided then there is api to extract the remaining two from whichever is implemented.

Default implementation will not provide enough flexibility to define which one will be the default.

Reply:
But if we have hardware acceleration, then you want to *replace* the function, which means you need a trait to override. I'm still uncertain I see the usefulness of these standing functions. The standing functions here are static, you can't change them, and the manufacturer won't be able to change them either.

Reply:
> But if we have hardware acceleration, then you want to *replace* the function, which means you need a trait to override. I'm still uncertain I see the usefulness of these standing functions. The standing functions here are static, you can't change them, and the manufacturer won't be able to change them either.

So, we strictly do not need them, they are utility helpers. They are meant to be used by `DRBG` trait implementors at their will, but the trait user must use the copy exposed by the trait.

Reply:
I would also add this link, it is about the the apis are used, I hope it help, it is from the software drbg implementation we he have, I hope it helps: https://github.com/tpm-rs/tpm-rs/blob/9724497d1508485afe5b8c49eb38311d847c2d4c/drbg/src/hashdrbg.rs#L193